java/arcs/core/storage/util/AbstractProxyCallbackManager.kt

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage.util
+
+import arcs.core.crdt.CrdtData
+import arcs.core.crdt.CrdtOperation
+import arcs.core.storage.ProxyCallback
+import arcs.core.storage.ProxyMessage
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * Thread-safe base manager for [ProxyCallback]s.
+ *
+ * Implementations can define how they wish their callback tokens to be generated.
+ */
+abstract class AbstractProxyCallbackManager<Data : CrdtData, Op : CrdtOperation, ConsumerData> {
+    private val mutex = Mutex()
+    /* internal */ val callbacks = mutableMapOf<Int, ProxyCallback<Data, Op, ConsumerData>>()
+
+    /**
+     * Generates a token [Int] to be assigned to an incoming [ProxyCallback].
+     */
+    abstract fun getNextToken(currentlyUsedTokens: Set<Int>): Int
+
+    /** Adds a [ProxyCallback] to the collection, and returns its token. */
+    fun register(proxyCallback: ProxyCallback<Data, Op, ConsumerData>): Int {
+        while (!mutex.tryLock()) { /* Wait. */ }
```

Comment:
Why this instead of `withLock`?

Reply:
This was already in the code before (looks new here, since it was refactored into this class from the original ProxyCallbackManager).

The reason I couldn't use `withLock` is because that's a suspend function, and I didn't want to make register a suspend fun itself.